update_packets.cc

```diff
@@ -29,10 +29,39 @@ namespace {
 
 void do_nonthermal_predeposit(Packet &pkt, const int nts, const double t2) {
   double en_deposited = pkt.e_cmf;
+  const double ts = pkt.prop_time;
 
-  if constexpr (INSTANT_PARTICLE_DEPOSITION) {
+  if constexpr (PARTICLE_THERMALISATION_SCHEME == ThermalisationScheme::INSTANT) {
     // absorption happens
     pkt.type = TYPE_NTLEPTON;
+  } else if constexpr (PARTICLE_THERMALISATION_SCHEME == ThermalisationScheme::BARNES_GLOBAL ||
+             PARTICLE_THERMALISATION_SCHEME == ThermalisationScheme::BARNES_LOCAL) {
+    const double t_0 = grid::get_t_model();
+    double rho_0 = 0.;
+    const double endot_per_rho = (pkt.pellet_decaytype == decay::DECAYTYPE_ALPHA) ? 5.e11 * MEV : 4.e10 * MEV;
+    if (PARTICLE_THERMALISATION_SCHEME == ThermalisationScheme::BARNES_GLOBAL) {
+      double V_0 = 0.;
+      for (int nonemptymgi = 0; nonemptymgi < grid::get_nonempty_npts_model(); nonemptymgi++) {
+        const int mgi = grid::get_mgi_of_nonemptymgi(nonemptymgi);
+        V_0 += grid::get_modelcell_assocvolume_tmin(mgi);
+      }
```

Comment:
I see two possible mistakes here.
- This volume does not include the volume of empty cells. Perhaps you want something like 4/3 * pi * (globals::vmax *globals::tmin)^3 for a spherical approximation of volume? or (2 * vmax * globals::tmin)^3 for the full cube.
- V_0 is the volume of non-empty cells at tmin (the simulation start time). Then rho_0 is the density at tmin. But in the expression for C, why is rho_0 multiplied by t_model^3? I can't follow the logic.

Reply:
This also looks totally unfamiliar to me from when I calculated the Barnes thermalisation efficiencies for plotting in artistools. Why are they so different?

```python
        ejecta_ke = (0.5 * (dfmodel["mass_g"] / 1000.0) * (dfmodel["vel_r_mid"] / 100.0) ** 2).sum()
        # velocity derived from ejecta kinetic energy to match Barnes et al. (2016) Section 2.1
        ejecta_v = np.sqrt(2 * ejecta_ke / (model_mass_grams * 1e-3))
        v2 = ejecta_v / (0.2 * 299792458)
        print(f"  Barnes average ejecta velocity: {ejecta_v / 299792458:.2f}c")
        m5 = model_mass_grams / (5e-3 * 1.989e33)  # M / (5e-3 Msun)

        e0_beta_mev = 0.5
        # Barnes et al (2016) equation 20
        t_ineff_beta = 7.4 * (e0_beta_mev / 0.5) ** -0.5 * m5**0.5 * (v2 ** (-3.0 / 2))
        # Barnes et al (2016) equation 32
        barnes_f_beta = math.log(1 + 2 * (t / t_ineff_beta) ** 2) / (2 * (t / t_ineff_beta) ** 2)
        
        e0_alpha_mev = 6.0
        # Barnes et al (2016) equation 25 times equation 16 for t_peak
        t_ineff_alpha = 4.3 * 1.8 * (e0_alpha_mev / 6.0) ** -0.5 * m5**0.5 * (v2 ** (-3.0 / 2))
        # Barnes et al (2016) equation 32
        barnes_f_alpha =math.log(1 + 2 * (t / t_ineff_alpha) ** 2) / (2 * (t / t_ineff_alpha) ** 2)
```

Reply:
You're right in both points. In the first case, I think I have to choose the spherical approximation (which is used by Barnes), in the second point that's just a mistake, I need a reference time / density where which has to be consistent.

Reply:
> I see two possible mistakes here.
> 
>     * This volume does not include the volume of empty cells. Perhaps you want something like 4/3 * pi * (globals::vmax *globals::tmin)^3 for a spherical approximation of volume? or (2 * vmax * globals::tmin)^3 for the full cube.
> 
>     * V_0 is the volume of non-empty cells at tmin (the simulation start time). Then rho_0 is the density at tmin. But in the expression for C, why is rho_0 multiplied by t_model^3? I can't follow the logic.

You're right in both points. In the first case, I think I have to choose the spherical approximation (which is used by Barnes), in the second point that's just a mistake, I need a reference time / density where which has to be consistent.

Reply:
> This also looks totally unfamiliar to me from when I calculated the Barnes thermalisation efficiencies for plotting in artistools. Why are they so different?
> 
> ```python
>         ejecta_ke = (0.5 * (dfmodel["mass_g"] / 1000.0) * (dfmodel["vel_r_mid"] / 100.0) ** 2).sum()
>         # velocity derived from ejecta kinetic energy to match Barnes et al. (2016) Section 2.1
>         ejecta_v = np.sqrt(2 * ejecta_ke / (model_mass_grams * 1e-3))
>         v2 = ejecta_v / (0.2 * 299792458)
>         print(f"  Barnes average ejecta velocity: {ejecta_v / 299792458:.2f}c")
>         m5 = model_mass_grams / (5e-3 * 1.989e33)  # M / (5e-3 Msun)
> 
>         e0_beta_mev = 0.5
>         # Barnes et al (2016) equation 20
>         t_ineff_beta = 7.4 * (e0_beta_mev / 0.5) ** -0.5 * m5**0.5 * (v2 ** (-3.0 / 2))
>         # Barnes et al (2016) equation 32
>         barnes_f_beta = math.log(1 + 2 * (t / t_ineff_beta) ** 2) / (2 * (t / t_ineff_beta) ** 2)
>         
>         e0_alpha_mev = 6.0
>         # Barnes et al (2016) equation 25 times equation 16 for t_peak
>         t_ineff_alpha = 4.3 * 1.8 * (e0_alpha_mev / 6.0) ** -0.5 * m5**0.5 * (v2 ** (-3.0 / 2))
>         # Barnes et al (2016) equation 32
>         barnes_f_alpha =math.log(1 + 2 * (t / t_ineff_alpha) ** 2) / (2 * (t / t_ineff_alpha) ** 2)
> ```

Those equations are for specific values of M_ej etc., right? I thought it would be better to use the general formula which leaves M_ej, v_ej as parameters. But I still tried to use the basic assumption of Barnes of a uniform sphere